submitted

flysputnik17 opened the pull request:
.

flysputnik17 committed:
submitted

.

gennady-bars commented on src/components/Api.js:
```diff
@@ -0,0 +1,125 @@
+      method: "GET",
+      headers: this._headers,
+    })
+      .then(this._checkResponse)
```
**Could be improved**

You can make a special method (function ) for fetching and checking responses **not to duplicate** it in every request:

```js

  _request(url, options) {
    return fetch(url, options).then(this._checkResponse)
  }

```

And now you can safely replace `fetch` with `this._request` and remove repeated checking. And you don’t need to change anything else.

gennady-bars commented on src/components/Api.js:
```diff
@@ -0,0 +1,125 @@
+      }),
+    })
+      .then(this._checkResponse)
+      .catch((error) => {
```
It’s better not to use `catch` in `Api` because you need to `catch` possible errors at the very **end** of any request and this end is in `index.js`  after all blocks `then`. And in real projects blocks `catch` usually open error popups and so on.

flysputnik17 replied:
i removed all the .catch in the Api.js and add them in the index.js after all .then in every api call i make

gennady-bars commented on src/components/PopupWithConfirmation.js:
```diff
@@ -0,0 +1,42 @@
+  the submit event
+  */
+  setSubmitAction(handler) {
+    this._handleSubmit = handler;
```
Nice way of function updating

gennady-bars commented on src/components/PopupWithConfirmation.js:
```diff
@@ -0,0 +1,42 @@
+    this._popupForm.addEventListener("submit", (evt) => {
+      evt.preventDefault();
+      this._handleSubmit(); //its called form here since we wating for a submit event
+      super.close();
```
You need to close popups only in blocks `then` after a **successful response** from the server so that when there is an error it would be opened (the inputs will not be cleared and the user could send the data again to the server) and if you change the button text you could see the change as well

**Please, correct it everywhere**

flysputnik17 replied:
i removed the super.close() from the EventListener and now the popup will be closed form the .then after the api call in the index,js 
api
    .upDateAvater(avatar.link)
    .then((res) => {
      avatarImage.src = res.avatar;
      avatarPopup.close();
      avatarUpdateValidator.resetValidation();
    })

gennady-bars commented on src/components/PopupWithForm.js:
```diff
@@ -0,0 +1,50 @@
+    });
+  }
+
+  setLoading(loading) {
+    if (loading) {
+      this._submitButton.textContent = "Saving...";
+    } else {
+      this._submitButton.textContent = "Save";
```
**Could be improved**

It’s better to make method `renderLoading` inside class `PopupWithForm`. You should find the button once in the constructor and change the text:  

```js
  constructor({ popupSelector, handleSubmit }) {
    super(popupSelector);
    // fix the initial button text only once in the constructor
    this._submitBtnText = this._submitBtn.textContent
  }
  // add 2 params: isLoading and loadingText with a default text
  renderLoading(isLoading, loadingText='Saving...') {
    if (isLoading) {
      this._submitBtn.textContent = loadingText;
    } else {
	// here we return back the initial text. So, you don’t need to bother yourself about it
      this._submitBtn.textContent = this._submitBtnText;
    }
  }

```

gennady-bars commented on src/components/Section.js:
```diff
@@ -0,0 +1,25 @@
+  constructor({ items = [], renderer }, cardSelector) {
+    this._items = items;
+    this._renderer = renderer;
+    this._cardSelector = document.querySelector(cardSelector);
```
`_cardSelector`

Word `selector` means a **string** that is used for searching elements in the DOM

flysputnik17 replied:
i changed the cardSelector to cardElement

gennady-bars commented on src/components/Section.js:
```diff
@@ -0,0 +1,25 @@
+
+  renderItems() {
+    this._items.forEach(({ name, link, _id }) => {
+      const element = {
+        name: name,
+        link: link,
+        _id: _id, // Assuming _id is the cardId
```
No need to create the same object

```js

    this._items.forEach((element) => {
      const newItem = this._renderer(element);
```

flysputnik17 replied:
this is really cool i changed it and its look more cleaner strange i didn't notice it when i wrote this in the first place

gennady-bars commented on src/components/UserInfo.js:
```diff
@@ -0,0 +1,25 @@
+    };
+  }
+
+  setUserInfo(name, description) {
+    this._name.textContent = name;
+    this._job.textContent = description;
+  }
+
+  loadUserInfo(name, description, avatar) {
+    this._name.textContent = name;
+    this._job.textContent = description;
+    this._avatar.src = avatar;
+  }
```
`loadUserInfo` duplicates method `setUserInfo`. Please, remove it or reuse it here

flysputnik17 replied:
i tried to merge those to into one function but when i made it i noticed i start to get problem with the avatar img every time i update the name and description after submitting the new inputs the avatar image will disappear and only after refreshing the web page the avatar image will appear normally again so i did this :
 setUserInfo(name, description) {
    this._name.textContent = name;
    this._job.textContent = description;
  }

  loadUserInfo(name, description, avatar) {
    this.setUserInfo(name, description);
    this._avatar.src = avatar;
  }

gennady-bars commented on src/pages/index.js:
```diff
@@ -0,0 +1,201 @@
+import Api from "../components/Api.js";
+import PopupWithConfirmation from "../components/PopupWithConfirmation.js";
+
+const editFormValidator = new FormValidator(config, profileForm); //creating a new var for the edit modal using the FormValidator class
+const addCardFormValidator = new FormValidator(config, cardForm);
```
**Could be improved**

You can **universally** create instances of validators for all forms in the project storing them inside one object `formValidators`.  And then you can take any validator using attribute `name` of the form where you need to disable/enable the submit button or remove errors.

```js
// define an object for storing validators
const formValidators = {}

const enableValidation = (config) => {
  const formList = Array.from(document.querySelectorAll(config.formSelector))
  formList.forEach((formElement) => {
    const validator = new FormValidator(config, formElement)
    // here you get the name of the form
    const formName = formElement.getAttribute('name')

   // here you store the validator using the `name` of the form
    formValidators[formName] = validator;
   validator.enableValidation();
  });
};

enableValidation(config);

```

And now you can use them for disabling buttons or clearing errors:

```js

formValidators[ profileForm.getAttribute('name') ].resetValidation()

// or you can use a string – the name of the form (you know it from `index.html`)

formValidators['profile-form'].resetValidation()

```

gennady-bars commented on src/components/Api.js:
```diff
@@ -0,0 +1,125 @@
+export default class Api {
+  constructor() {
+    this._baseUrl = "https://around-api.en.tripleten-services.com/v1";
+    this._headers = {
+      authorization: "bfb869e8-08ec-4b67-8cc1-518f5a35ed9e",
+      "Content-Type": "application/json",
+    };
```
According to the task you need to pass full `headers` and `baseUrl` into the constructor of `Api`. See here https://disk.yandex.ru/i/J93dbJp7eypXKQ . 

You shouldn't hardcode them inside the class

flysputnik17 replied:
fixed it

gennady-bars commented on src/pages/index.js:
```diff
@@ -0,0 +1,201 @@
+  if (card.isLiked) {
+    api.dislikeCard(card._id).then((res) => {
+      card.setLike(res.isLiked);
+    });
```
You need to `catch` possible errors at the very end of any server request.

**Please, correct it everywhere**

flysputnik17 replied:
fixed it

gennady-bars commented on src/pages/index.js:
```diff
@@ -0,0 +1,201 @@
+function upDateAvatar(avatar) {
+  avatarPopup.setLoading(true);
+  api.upDateAvater(avatar.link).then((res) => {
+    avatarImage.src = res.avatar;
```
Class `UserInfo` should set the entire user’s data (including the **avatar**)

flysputnik17 replied:
i factored it 

function upDateAvatar(avatar) {
  avatarPopup.setLoading(true);
  api
    .upDateAvater(avatar.link)
    .then((res) => {
      avatarImage.src = res.avatar;
      avatarPopup.close();
      avatarUpdateValidator.resetValidation();
    })
    .catch((error) => {
      console.error("Error in upDateAvatar:", error);
      return Promise.reject(error);
    })
    .finally(() => {
      avatarPopup.setLoading(false);
      avatarUpdateValidator.resetValidation();
    });
}

gennady-bars commented on src/pages/index.js:
```diff
@@ -0,0 +1,201 @@
+    avatarImage.src = res.avatar;
+  });
+  avatarUpdateValidator.resetValidation();
+  avatarPopup.setLoading(false);
```
You need to return the **default** button text only in blocks `finally` so that the button could get the text back in any case: `success` or `failure`.

**Please, correct it everywhere**

flysputnik17 replied:
fixed it

gennady-bars commented on src/pages/index.js:
```diff
@@ -0,0 +1,201 @@
+  });
+  avatarUpdateValidator.resetValidation();
+  avatarPopup.setLoading(false);
+  avatarPopup.close();
```
You need to close popups only in blocks `then` after a **successful response** from the server so that when there is an error it would be opened (the inputs will not be cleared and the user could send the data again to the server) and if you change the button text you could see the change as well

**Please, correct it everywhere**

flysputnik17 replied:
fixed it

gennady-bars commented on src/pages/index.js:
```diff
@@ -0,0 +1,201 @@
+}
+
+function upDateAvatar(avatar) {
+  avatarPopup.setLoading(true);
```
**Could be improved**

If it’s interesting for you here is how we can make a universal function for handling any submit. We can get rid of such duplicating as **loading effect**, **closing** and **catching errors**


```js

// You can make a universal function that accepts a request, popup instance and optional loading text  
function handleSubmit(request, popupInstance, loadingText = "Saving...") {

  // here we change the button text
  popupInstance.renderLoading(true, loadingText);
  request()
    .then(() => {
      // We need to close only in `then`
      popupInstance.close()
    })
      // we need to catch possible errors
      // console.error is used to handle errors if you don’t have any other ways for that
     .catch(console.error)
    // in `finally` we need to return the initial button text back in any case
    .finally(() => {
      popupInstance.renderLoading(false);
    });
}



```


You need to pass the **request** into the call of `handleSubmit` like that:



```js

// here is an example of the profile form handling
function handleProfileFormSubmit(inputValues) {
  // we create a function that returns a promise
  function makeRequest() {
    // `return` lets us use a promise chain `then, catch, finally` inside `handleSubmit`
    return api.editProfile(inputValues).then((userData) => {
      userInfo.setUserInfo(userData)
    });
  }
  // Here we call the function passing the request, popup instance and if we need some other loading text we can pass it as the 3rd argument
  handleSubmit(makeRequest, profilePopup);
}



```

gennady-bars commented on src/pages/index.js:
```diff
@@ -0,0 +1,201 @@
+
+//////////////////////////// modal events ///////////////////
+editButton.addEventListener("click", function () {
+  api.getUserInfo().then((userData) => {
```
` api.getUserInfo()` shouldn't be called when you open the profile because you have method `userInfo.getUserInfo` for getting the data difrectly from the page without waiting

flysputnik17 replied:
changed it to : 

editButton.addEventListener("click", function () {
  nameInput.value = profileName.textContent;
  jobInput.value = profileJob.textContent;
  newEditPopup.open();
});


i dont need the userInfo.getUserInfo since in the constants,js i query selecting the profileName and profileJob from the DOM

gennady-bars commented on src/pages/index.js:
```diff
@@ -0,0 +1,201 @@
+  api.upDateAvater(avatar.link).then((res) => {
+    avatarImage.src = res.avatar;
+  });
+  avatarUpdateValidator.resetValidation();
```
`  avatarUpdateValidator.resetValidation();
`

This should be done in block `then` where you'll close the modal because closing will be in `then` as well

flysputnik17 replied:
fixed it

gennady-bars reviewed (commented):
Dear Pavel , (**Please, expand the general comment ↓**)

Good job but there are some remarks:



* New cards should be added only with `prepend` to the start of the section otherwise the user doesn’t see the new card and may think that it hasn’t been created
* According to the task you need to change submit buttons text (`Saving…`) while waiting for the server response. And the user should see it.
* It’s better not to use `catch` in `Api` because you need to `catch` possible errors at the very **end** of any request and this end is in `index.js`  after all blocks `then`. And in real projects blocks `catch` usually open error popups and so on.
* You need to close popups only in blocks `then` after a **successful response** from the server so that when there is an error it would be opened (the inputs will not be cleared and the user could send the data again to the server) and if you change the button text you could see the change as well
* Word `selector` means a **string** that is used for searching elements in the DOM
* You need to `catch` possible errors at the very end of any server request.
* Class `UserInfo` should set the entire user’s data (including the **avatar**)
* You need to return the **default** button text only in blocks `finally` so that the button could get the text back in any case: `success` or `failure`.
* ` api.getUserInfo()` shouldn't be called when you open the profile because you have method `userInfo.getUserInfo` for getting the data difrectly from the page without waiting


Please, correct everything and your project will be accepted.

Good luck with the refactoring

flysputnik17 committed:
refactoring

refactoring after first submission

flysputnik17 committed:
popup not closing

popup will not close and the input will not reset if there is an error

flysputnik17 committed:
fixed button toggle

the button of the submit form was active even after incorrect input or after a submission so now every time the form popup will open the button of the form will be inactive and if the user put incorrect values the popup will stay open and the inputs will not reset

gennady-bars commented on src/components/Section.js:
```diff
@@ -0,0 +1,18 @@
+  constructor({ items = [], renderer }, cardElement) {
+    this._items = items;
+    this._renderer = renderer;
+    this._cardSelector = document.querySelector(cardElement);
```
`_cardSelector`


You haven't corrected it to `_cardsContainer` or just `_container`

flysputnik17 replied:
sorry i thought the cardSelector is the wrong one 
i fixed it

gennady-bars commented on src/pages/index.js:
```diff
@@ -0,0 +1,264 @@
+    })
+    .catch((error) => {
+      console.error("Error in addNewCard:", error);
+      return Promise.reject(error);
```
`Promise.reject` is not needed in block `catch` because it **throws** the error again outside the chain. The final `catch` should keep the error inside

**Please, correct it everywhere**

flysputnik17 replied:
fixed it i see what you mean 
for example if i get error 400 i will get it twice

gennady-bars commented on src/pages/index.js:
```diff
@@ -0,0 +1,264 @@
+  api
+    .upDateAvater(avatar.link)
+    .then((res) => {
+      avatarImage.src = res.avatar;
```
Class `UserInfo` should set the entire user’s data (including the **avatar**)

flysputnik17 replied:
fixed it now the userInfo updating everything

gennady-bars commented on src/pages/index.js:
```diff
@@ -0,0 +1,264 @@
+// });
+
+editButton.addEventListener("click", function () {
+  nameInput.value = profileName.textContent;
+  jobInput.value = profileJob.textContent;
```
`textContent` shouldn't be used in `index.js` anymore. You have `UserInfo` for getting the user info

flysputnik17 replied:
i fixed it

gennady-bars commented on src/pages/index.js:
```diff
@@ -0,0 +1,264 @@
+  //and in the .then method after we get a response we calling the setUserInfo of the UserInfo class
+  //and passing it the userData as a propery{name,about and avatar} are proporty of the body we get from the
+  //API and in the setUserInfo we assinig the name and about to the name and description textContent
+});
```
You need to `catch` possible errors at the very end of any server request.

**Please, correct it everywhere**

flysputnik17 replied:
fixed it

gennady-bars commented on src/pages/index.js:
```diff
@@ -0,0 +1,264 @@
+// });
+
+editButton.addEventListener("click", function () {
+  nameInput.value = profileName.textContent;
```
**Could be improved**

Please, call `resetValidation` for the profile form to clear possible errors

flysputnik17 replied:
fixed it

gennady-bars reviewed (commented):
Everything is almost fine but:


* You haven't corrected something https://github.com/flysputnik17/se_project_aroundtheus/pull/1#discussion_r1492777947 
* `Promise.reject` is not needed in block `catch` because it **throws** the error again outside the chain. The final `catch` should keep the error inside
* Class `UserInfo` should set the entire user’s data (including the **avatar**)
* `textContent` shouldn't be used in `index.js` anymore. You have `UserInfo` for getting the user info
* You need to `catch` possible errors at the very end of any server request.

flysputnik17 committed:
fixing code review

gennady-bars reviewed (approved):
You’ve done a great job.

Your project has been accepted. 

Good luck
